playout/libretime_playout/liquidsoap/entrypoint.py

```diff
@@ -1,64 +1,82 @@
-import os
-import sys
-import time
+from enum import Enum
 from pathlib import Path
-from typing import Optional
+from typing import Optional, Tuple
 
-from libretime_api_client.v1 import ApiClient as LegacyClient
-from loguru import logger
+from jinja2 import Template
+from libretime_shared.config import StreamIcecastOutput
+from pydantic import BaseModel
 
+from ..config import Config
 
-def generate_liquidsoap_config(ss, log_filepath: Optional[Path]):
-    data = ss["msg"]
-    fh = open("/etc/libretime/liquidsoap.cfg", "w")
-    fh.write("################################################\n")
-    fh.write("# THIS FILE IS AUTO GENERATED. DO NOT CHANGE!! #\n")
-    fh.write("################################################\n")
-    fh.write("# The ignore() lines are to squash unused variable warnings\n")
+here = Path(__file__).parent
 
-    for key, value in data.items():
-        try:
-            if not "port" in key and not "bitrate" in key:  # Stupid hack
-                raise ValueError()
-            str_buffer = f"{key} = {int(value)}\n"
-        except ValueError:
-            try:  # Is it a boolean?
-                if value == "true" or value == "false":
-                    str_buffer = f"{key} = {value.lower()}\n"
-                else:
-                    raise ValueError()  # Just drop into the except below
-            except:  # Everything else is a string
-                str_buffer = f'{key} = "{value}"\n'
+entrypoint_template_path = here / "entrypoint.liq.j2"
+entrypoint_template = Template(
+    entrypoint_template_path.read_text(encoding="utf-8"),
+    keep_trailing_newline=True,
+)
 
-        fh.write(str_buffer)
-        # ignore squashes unused variable errors from Liquidsoap
-        fh.write("ignore(%s)\n" % key)
+# Liquidsoap has 4 hardcoded output stream set of variables, so we need to
+# fill the missing stream outputs with placeholders so Liquidsoap does
+# not fail with missing variables in the entrypoint.
+_stream_output_placeholder = StreamIcecastOutput(
+    enabled=False,
+    mount="",
+    source_password="",
+    audio=dict(format="ogg", bitrate=256),
+)
 
-    auth_path = os.path.dirname(os.path.realpath(__file__))
-    log_file = log_filepath.resolve() if log_filepath is not None else ""
 
-    fh.write(f'log_file = "{log_file}"\n')
-    fh.write('auth_path = "%s/liquidsoap_auth.py"\n' % auth_path)
-    fh.close()
+class MessageFormatKind(int, Enum):
+    ARTIST_TITLE = 0
+    SHOW_ARTIST_TITLE = 1
+    RADIO_SHOW = 2
 
 
-def generate_entrypoint(log_filepath: Optional[Path]):
-    attempts = 0
-    max_attempts = 10
-    successful = False
+class StreamSettings(BaseModel):
+    station_name: str
+    input_fade_transition: float
+    message_format: MessageFormatKind
+    message_offline: str
+    source_input_main_streaming: bool
+    source_input_show_streaming: bool
+    source_schedule_streaming: bool
```

Comment:
Sharing those models across api and playout is not a recommended pattern, we will have to maintain models on both sides.

Reply:
Can they not go in `shared`?

Reply:
This would be great, but I am afraid of few things :
-  If we want to modify a model in one service, we have to apply the change to the other of the services using it. That is doable but commits might not represent what they supposed to changes as many other parts will be edited. Being able to modify an api and only map old => new field on model setup seem easier. This is a really important in micro services architecture, but I think it can be applied in our case, as we shared if for both the api and playout.

- Sharing logic might be a good idea for small things, but I intend to use some of those models to build advanced features that are tightly integrated in the services using it, pushing everything in the shared lib will just recreate a api-client like library were we had the playout code is spread in multiple packages.

We might be able to share those models as data structure only and inherit the classes to write some extra features, but we will have to be careful not to make it hard to use in the long run.

Also, maybe I am overthinking this and we can share it and see how it goes.

Reply:
Maybe we add it to shared for now and if things expand or get more complicated we can refactor things out?

Reply:
Ok !

Reply:
Back to this, I think we are already hitting a too complex scenario, for example the api preferences. 

For example:
https://github.com/jooola/libretime/blob/4b17894c6e98ff76fabebdc4987ef7bb09d1f4ec/api/libretime_api/core/models/preference.py#L17-L29

will be extend really soon, when we will uncomment fields. Playout don't need some of the extra fields but with the validation system, the values will still be required as they aren't marked as optional.

So there is illusion of duplication for now, but both models will diverge really soon. I don't think we should merge this into the shared lib.

In addition preferences models will be used all over the place, maybe we should not add complexity from the start.

If we want to make sure the models are in sync for the few fields they share, we should probably create tests instead.

Reply:
Ok, sounds reasonable to keep them separate then